File: longest_substring_with_at_most_k_distinct_characters/LongestSubstringWithAtMostKDistinctCharacters.java
class LongestSubstringWithAtMostKDistinctCharacters {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        if (s == null || s.length() == 0 || k < 1) {
            return 0;
        }
        
        Map<Character, int[]> table = new HashMap<>();
        int slow = 0;
        int maxLen = 0;
        for (int fast = 0; fast < s.length(); fast++) {
            char ch = s.charAt(fast);
            if (!table.containsKey(ch)) {
                table.put(ch, new int[]{0});
            }
            table.get(ch)[0]++;
            if (table.size() <= k) {
                maxLen = Math.max(maxLen, fast - slow + 1);
                continue;
            }
            
            while (table.size() > k && --table.get(s.charAt(slow++))[0] > 0);
            table.remove(s.charAt(slow - 1));
        }
        
        return maxLen;
    }
}

File: longest_substring_with_at_most_k_distinct_characters/Longest_Substring_With_At_Most_K_Distinct_Characters.cpp
class Longest_Substring_With_At_Most_K_Distinct_Characters {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        if (s.empty() || k < 1) {
            return 0;
        }

        unordered_map<char, int> table;
        int slow = 0;
        int max_len = 0;
        for (int fast = 0; fast < s.size(); fast++) {
            table[s[fast]]++;
            if (table.size() <= k) {
                max_len = max(max_len, fast - slow + 1);
                continue;
            }

            while (table.size() > k && --table[s[slow++]] > 0);
            table.erase(s[slow - 1]);
        }

        return max_len;
    }
};